src/audio/src/src_hifi4.c

```diff
@@ -0,0 +1,559 @@
+
+		/* Reset coefficient pointer and clear accumulator */
+		coefp = (ae_f16x4 *)cp;
+		a0 = AE_ZERO64();
+
+		/* Compute FIR filter for current channel with four
+		 * taps per every loop iteration. Data is read from
+		 * interleaved buffer with stride of channels count.
+		 */
+		for (i = 0; i < taps_div_4; i++) {
+			/* Load four coefficients */
+			AE_LA16X4_IP(coef4, u, coefp);
+
+			/* Load two data samples, place to high and
+			 * low of data2.
+			 */
+			AE_L32_XC(d0, dp0, inc);
+			AE_L32_XC(d1, dp0, inc);
+			data2 = AE_SEL32_LL(d0, d1);
+
+			/* Accumulate
+			 * data2_h * coef4_3 + data2_l* coef4_2.
+			 * The data is 32 bits Q1.31 and coefficient 16 bits
+			 * Q1.15. The accumulator is Q17.47.
+			 */
+			AE_MULAAFD32X16_H3_L2(a0, data2, coef4);
+
+			/* Repeat with next two samples */
+			AE_L32_XC(d0, dp0, inc);
+			AE_L32_XC(d1, dp0, inc);
+			data2 = AE_SEL32_LL(d0, d1);
+
+			/* Accumulate
+			 * data2_h * coef4_1 + data2_l * coef4_0.
+			 */
+			AE_MULAAFD32X16_H1_L0(a0, data2, coef4);
+		}
+
+		/* Scale FIR output with right shifts, round/saturate Q17.47
+		 * to Q1.31, and store 32 bit output. Advance write
+		 * pointer to next sample.
+		 */
+		AE_S32_L_XP(AE_ROUND32F48SSYM(AE_SRAA64(a0, shift)), wp,
+			    sizeof(int32_t));
+	}
+}
+
+#else /* 32bit coefficients version */
+
+static inline void fir_filter(ae_f32 *rp, const void *cp, ae_f32 *wp0,
+			      const int taps_div_4, const int shift,
+			      const int nch)
+{
+	/* This function uses
+	 * 6x 64 bit registers
+	 * 3x integers
+	 * 5x address pointers,
+	 */
+	ae_f64 a0;
+	ae_f64 a1;
+	ae_f24x2 data2 = AE_ZERO24();
+	ae_f24x2 coef2 = AE_ZERO24();
+	ae_f24x2 d0 = AE_ZERO24();
+	ae_f24x2 d1 = AE_ZERO24();
+	ae_f24x2 *coefp;
+	ae_f24x2 *dp;
+	ae_f24 *dp1;
+	ae_f24 *dp0;
+	int i;
+	int j;
+	ae_f32 *wp = wp0;
+	const int inc = nch * sizeof(int32_t);
+
+	if (nch == 2) {
+		/* Move data pointer back by one sample to start from right
+		 * channel sample. Discard read value p0.
+		 */
+		dp = (ae_f24x2 *)rp;
+		AE_L32F24_XC(d0, (ae_f24 *)dp, -sizeof(ae_f24));
+
+		/* Reset coefficient pointer and clear accumulator */
+		coefp = (ae_f24x2 *)cp;
+		a0 = AE_ZERO64();
+		a1 = AE_ZERO64();
+
+		/* Compute FIR filter for current channel with four
+		 * taps per every loop iteration.  Two coefficients
+		 * are loaded simultaneously. Data is read
+		 * from interleaved buffer with stride of channels
+		 * count.
+		 */
+		for (i = 0; i < taps_div_4; i++) {
+			/* Load two coefficients. Coef2_h contains tap *coefp
+			 * and coef2_l contains the next tap.
+			 */
+			/* TODO: Ensure coefficients are 64 bits aligned */
+			AE_L32X2F24_IP(coef2, coefp, sizeof(ae_f24x2));
+
+			/* Load two data samples from two channels */
+			AE_L32X2F24_XC(d0, dp, inc); /* r0, l0 */
+			AE_L32X2F24_XC(d1, dp, inc); /* r1, l1 */
+
+			/* Select to d0 successive left channel samples, to d1
+			 * successive right channel samples. Then Accumulate
+			 * to a0 and a1
```

Comment:
are these all "to" or should some of them be "two?" Also in other similar comments in the file